# Table of Contents
## 01. Importing Libraries
## 02. Importing Data
## 03. Exercise 4.8 Task
## 04. Exporting Data

## 01. Import Libraries

In [2]:
import pandas as pd
import numpy as np
import os

## 02. Import Data

In [3]:
path = r'/Users/lianabulte/Career Foundry/2023 02 Instacart Basket Analysis'

In [4]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
# verifying shape of imported data
ords_prods_merge.shape

(30328763, 18)

In [7]:
# reviewing columns of imported data
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2398795        1             2                   3                  7   
1    473747        1             3                   3                 12   
2   2254736        1             4                   4                  7   
3    431534        1             5                   4                 15   
4   3367565        1             6                   2                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    21.0         196                  1          1   
2                    29.0         196                  1          1   
3                    28.0         196                  1          1   
4                    19.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  
0  Regularly busy  Slowest days        Average orders  
1  Regularly busy  Slowest days           Most orders  
2      Least busy  Slowest days        Average orders  
3      Least busy  Slowest days           Most orders  
4  Regularly busy  Regular days        Average orders

## 03. Exercise 4.8 Task questions

#### 2. Find the aggregated mean of the 'order_number' column grouped by 'department_id'

In [8]:
# creating group by function to group order_number data by department_id
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                16.559358
2                18.413176
3                18.279600
4                18.915890
5                16.497751
6                17.609390
7                18.303975
8                16.383301
9                17.022963
10               21.227447
11               17.311768
12               16.953613
13               17.727470
14               17.898097
15               17.290123
16               18.757796
17               16.808683
18               20.336389
19               18.258620
20               17.578959
21               23.644477

3. Analyze the results - the data in the subset provided a smaller list of departments with their corresponding means. By applying the above code to the full dataset we are able to see the means changes as well as view the full list of departments. For example, the subset had department 17 with a mean of 12.64, while in the full dataset this mean has risen to 16.80 indicating that the items in this department are ordered more frequently that intitially assumed.

#### 4. Creating a loyalty flag for existing customers

In [9]:
# creating a new column 'max_order' to generate number of orders by user
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [10]:
# validate that new column appears in dataset
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2398795        1             2                   3                  7   
1    473747        1             3                   3                 12   
2   2254736        1             4                   4                  7   
3    431534        1             5                   4                 15   
4   3367565        1             6                   2                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    21.0         196                  1          1   
2                    29.0         196                  1          1   
3                    28.0         196                  1          1   
4                    19.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Slowest days        Average orders         10  
1  Regularly busy  Slowest days           Most orders         10  
2      Least busy  Slowest days        Average orders         10  
3      Least busy  Slowest days           Most orders         10  
4  Regularly busy  Regular days        Average orders         10

Creating loc for 'loyalty flag'

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# verify totals per flag group
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15081691
Loyal customer      10095381
New customer         5151691
Name: loyalty_flag, dtype: int64

In [15]:
# re-verifying shape of dataset after loc
ords_prods_merge.shape

(30328763, 20)

2 additional columns of 'max_orders' and 'loyalty_flag' added to initial 18 columns at the beginning, this aligns 

In [16]:
# re-validate that new columns appears in dataset
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2398795        1             2                   3                  7   
1     473747        1             3                   3                 12   
2    2254736        1             4                   4                  7   
3     431534        1             5                   4                 15   
4    3367565        1             6                   2                  7   
5     550135        1             7                   1                  9   
6    3108588        1             8                   1                 14   
7    2295261        1             9                   1                 16   
8    2550362        1            10                   4                  8   
9    2968173       15            15                   1                  9   
10   1870022       15            17                   2                 16   
11   1911383       15            18                   2                 11   
12   2715276       15            21                   1                  9   
13    487368       15            22                   1                 10   
14   2293453       19             2                   5                 14   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     21.0         196                  1          1   
2                     29.0         196                  1          1   
3                     28.0         196                  1          1   
4                     19.0         196                  1          1   
5                     20.0         196                  1          1   
6                     14.0         196                  2          1   
7                      0.0         196                  4          1   
8                     30.0         196                  1          1   
9                      7.0         196                  2          0   
10                     8.0         196                  6          1   
11                     7.0         196                  1          1   
12                     7.0         196                  1          1   
13                    14.0         196                  1          1   
14                     6.0         196                  3          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77              7     9.0   both  Mid-range product   
13         Soda        77              7     9.0   both  Mid-range product   
14         Soda        77              7     9.0   both  Mid-range product   

       busiest_day  busiest_days busiest_period_of_day  max_order  \
0   Regularly busy  Slowest days        Average orders         10   
1   Regularly busy  Slowest days           Most orders         10   
2       Least busy  Slowest days        Average orders         10   
3       Least busy  Slowest days           Most orders         10   
4   Regu

#### 5. Use loyalty flag to identify the prices of products purchased between loyalty groups

In [18]:
# creating group by function to group prices data by loyalty_flag
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.402162  1.0  99999.0
New customer      13.396333  1.0  99999.0
Regular customer  12.546842  1.0  99999.0

Based on the above statistics new customers spend more money. However, the max price for the dataset being 99,999 leads me to believe that there is a pricing error on an item that is most likely brinding the mean price up more than it should be. 

#### 6. Create a spending flag for each user based on the average price across all orders

1. If the mean of the prices of products purchased by a user is < than 10 = flag as 'Low spender'
2. If the mean of the prices of products purchased by a user is >= to 10 = flag as 'High spender'

In [19]:
# creating a new column 'avg_price' to generate mean price of products by user
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [20]:
# validate that new column appears in dataset
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2398795        1             2                   3                  7   
1     473747        1             3                   3                 12   
2    2254736        1             4                   4                  7   
3     431534        1             5                   4                 15   
4    3367565        1             6                   2                  7   
5     550135        1             7                   1                  9   
6    3108588        1             8                   1                 14   
7    2295261        1             9                   1                 16   
8    2550362        1            10                   4                  8   
9    2968173       15            15                   1                  9   
10   1870022       15            17                   2                 16   
11   1911383       15            18                   2                 11   
12   2715276       15            21                   1                  9   
13    487368       15            22                   1                 10   
14   2293453       19             2                   5                 14   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     21.0         196                  1          1   
2                     29.0         196                  1          1   
3                     28.0         196                  1          1   
4                     19.0         196                  1          1   
5                     20.0         196                  1          1   
6                     14.0         196                  2          1   
7                      0.0         196                  4          1   
8                     30.0         196                  1          1   
9                      7.0         196                  2          0   
10                     8.0         196                  6          1   
11                     7.0         196                  1          1   
12                     7.0         196                  1          1   
13                    14.0         196                  1          1   
14                     6.0         196                  3          0   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid-range product   
1          Soda  ...              7     9.0    both  Mid-range product   
2          Soda  ...              7     9.0    both  Mid-range product   
3          Soda  ...              7     9.0    both  Mid-range product   
4          Soda  ...              7     9.0    both  Mid-range product   
5          Soda  ...              7     9.0    both  Mid-range product   
6          Soda  ...              7     9.0    both  Mid-range product   
7          Soda  ...              7     9.0    both  Mid-range product   
8          Soda  ...              7     9.0    both  Mid-range product   
9          Soda  ...              7     9.0    both  Mid-range product   
10         Soda  ...              7     9.0    both  Mid-range product   
11         Soda  ...              7     9.0    both  Mid-range product   
12         Soda  ...              7     9.0    both  Mid-range product   
13         Soda  ...              7     9.0    both  Mid-range product   
14         Soda  ...              7     9.0    both  Mid-range product   

       busiest_day  busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Slowest days        Average orders        10   
1   Regularly busy  Slowest days           Most orders        10   
2       Least busy  Slowest days        Average orders        10   
3       Least busy  Slowest days           Most orders        10   
4   Regularly busy  Regular days        Average orders        10   
5   Regul

Creating loc for 'spending flag'

In [21]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [22]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [23]:
# verify totals per flag group
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     29730505
High spender      598258
Name: spending_flag, dtype: int64

In [24]:
# re-verifying shape of dataset after loc
ords_prods_merge.shape

(30328763, 22)

In [25]:
# re-validate that new columns appears in dataset
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2398795        1             2                   3                  7   
1     473747        1             3                   3                 12   
2    2254736        1             4                   4                  7   
3     431534        1             5                   4                 15   
4    3367565        1             6                   2                  7   
5     550135        1             7                   1                  9   
6    3108588        1             8                   1                 14   
7    2295261        1             9                   1                 16   
8    2550362        1            10                   4                  8   
9    2968173       15            15                   1                  9   
10   1870022       15            17                   2                 16   
11   1911383       15            18                   2                 11   
12   2715276       15            21                   1                  9   
13    487368       15            22                   1                 10   
14   2293453       19             2                   5                 14   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     21.0         196                  1          1   
2                     29.0         196                  1          1   
3                     28.0         196                  1          1   
4                     19.0         196                  1          1   
5                     20.0         196                  1          1   
6                     14.0         196                  2          1   
7                      0.0         196                  4          1   
8                     30.0         196                  1          1   
9                      7.0         196                  2          0   
10                     8.0         196                  6          1   
11                     7.0         196                  1          1   
12                     7.0         196                  1          1   
13                    14.0         196                  1          1   
14                     6.0         196                  3          0   

   product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0          Soda  ...     9.0    both  Mid-range product  Regularly busy   
1          Soda  ...     9.0    both  Mid-range product  Regularly busy   
2          Soda  ...     9.0    both  Mid-range product      Least busy   
3          Soda  ...     9.0    both  Mid-range product      Least busy   
4          Soda  ...     9.0    both  Mid-range product  Regularly busy   
5          Soda  ...     9.0    both  Mid-range product  Regularly busy   
6          Soda  ...     9.0    both  Mid-range product  Regularly busy   
7          Soda  ...     9.0    both  Mid-range product  Regularly busy   
8          Soda  ...     9.0    both  Mid-range product      Least busy   
9          Soda  ...     9.0    both  Mid-range product  Regularly busy   
10         Soda  ...     9.0    both  Mid-range product  Regularly busy   
11         Soda  ...     9.0    both  Mid-range product  Regularly busy   
12         Soda  ...     9.0    both  Mid-range product  Regularly busy   
13         Soda  ...     9.0    both  Mid-range product  Regularly busy   
14         Soda  ...     9.0    both  Mid-range product  Regularly busy   

    busiest_days busiest_period_of_day max_order      loyalty_flag  avg_price  \
0   Slowest days        Average orders        10      New customer   6.372222   
1   Slowest days           Most orders        10      New customer   6.372222   
2   Slowest days        Average orders        10      New customer   6.372222   
3   Slowest days           Most orders        10      New customer   6.372222

Average price and spending_flag aligned

#### 7. Create 'order frequency flag' that marks regularity of a user's ordering behaviour according to the median in the 'days since prior order' column

1. If the median of 'days since prior order' is > than 20, customer flag = 'non-frequent customer'
2. If the median of 'days since prior order' is > than 10 and <= 20, customer flag = 'regular customer'
3. If the median of 'days since prior order' is <= 10, customer flag = 'frequent customer'

In [26]:
# creating a new column 'median_prior_order' to generate order frequency by user
ords_prods_merge['median_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [27]:
# validate that new column appears in dataset
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2398795        1             2                   3                  7   
1     473747        1             3                   3                 12   
2    2254736        1             4                   4                  7   
3     431534        1             5                   4                 15   
4    3367565        1             6                   2                  7   
5     550135        1             7                   1                  9   
6    3108588        1             8                   1                 14   
7    2295261        1             9                   1                 16   
8    2550362        1            10                   4                  8   
9    2968173       15            15                   1                  9   
10   1870022       15            17                   2                 16   
11   1911383       15            18                   2                 11   
12   2715276       15            21                   1                  9   
13    487368       15            22                   1                 10   
14   2293453       19             2                   5                 14   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     21.0         196                  1          1   
2                     29.0         196                  1          1   
3                     28.0         196                  1          1   
4                     19.0         196                  1          1   
5                     20.0         196                  1          1   
6                     14.0         196                  2          1   
7                      0.0         196                  4          1   
8                     30.0         196                  1          1   
9                      7.0         196                  2          0   
10                     8.0         196                  6          1   
11                     7.0         196                  1          1   
12                     7.0         196                  1          1   
13                    14.0         196                  1          1   
14                     6.0         196                  3          0   

   product_name  ...  _merge    price_range_loc     busiest_day  busiest_days  \
0          Soda  ...    both  Mid-range product  Regularly busy  Slowest days   
1          Soda  ...    both  Mid-range product  Regularly busy  Slowest days   
2          Soda  ...    both  Mid-range product      Least busy  Slowest days   
3          Soda  ...    both  Mid-range product      Least busy  Slowest days   
4          Soda  ...    both  Mid-range product  Regularly busy  Regular days   
5          Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
6          Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
7          Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
8          Soda  ...    both  Mid-range product      Least busy  Slowest days   
9          Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
10         Soda  ...    both  Mid-range product  Regularly busy  Regular days   
11         Soda  ...    both  Mid-range product  Regularly busy  Regular days   
12         Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
13         Soda  ...    both  Mid-range product  Regularly busy  Busiest days   
14         Soda  ...    both  Mid-range product  Regularly busy  Regular days   

   busiest_period_of_day max_order      loyalty_flag avg_price  spending_flag  \
0         Average orders        10      New customer  6.372222    Low spender   
1            Most orders        10      New customer  6.372222    Low spender   
2         Average orders        10      New customer  6.372222

Creating loc for 'order_frequency_flag'

In [28]:
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['median_prior_order'] > 10) & (ords_prods_merge['median_prior_order'] <=20), 'order_frequency_flag'] = 'Regular customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [31]:
# verify totals per flag group
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        20675290
Regular customer          6594542
Non-frequent customer     3058931
Name: order_frequency_flag, dtype: int64

In [32]:
# re-verifying shape of dataset after loc
ords_prods_merge.shape

(30328763, 24)

In [33]:
# re-validate that new columns appears in dataset
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2398795        1             2                   3                  7   
1     473747        1             3                   3                 12   
2    2254736        1             4                   4                  7   
3     431534        1             5                   4                 15   
4    3367565        1             6                   2                  7   
5     550135        1             7                   1                  9   
6    3108588        1             8                   1                 14   
7    2295261        1             9                   1                 16   
8    2550362        1            10                   4                  8   
9    2968173       15            15                   1                  9   
10   1870022       15            17                   2                 16   
11   1911383       15            18                   2                 11   
12   2715276       15            21                   1                  9   
13    487368       15            22                   1                 10   
14   2293453       19             2                   5                 14   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     21.0         196                  1          1   
2                     29.0         196                  1          1   
3                     28.0         196                  1          1   
4                     19.0         196                  1          1   
5                     20.0         196                  1          1   
6                     14.0         196                  2          1   
7                      0.0         196                  4          1   
8                     30.0         196                  1          1   
9                      7.0         196                  2          0   
10                     8.0         196                  6          1   
11                     7.0         196                  1          1   
12                     7.0         196                  1          1   
13                    14.0         196                  1          1   
14                     6.0         196                  3          0   

   product_name  ...    price_range_loc     busiest_day  busiest_days  \
0          Soda  ...  Mid-range product  Regularly busy  Slowest days   
1          Soda  ...  Mid-range product  Regularly busy  Slowest days   
2          Soda  ...  Mid-range product      Least busy  Slowest days   
3          Soda  ...  Mid-range product      Least busy  Slowest days   
4          Soda  ...  Mid-range product  Regularly busy  Regular days   
5          Soda  ...  Mid-range product  Regularly busy  Busiest days   
6          Soda  ...  Mid-range product  Regularly busy  Busiest days   
7          Soda  ...  Mid-range product  Regularly busy  Busiest days   
8          Soda  ...  Mid-range product      Least busy  Slowest days   
9          Soda  ...  Mid-range product  Regularly busy  Busiest days   
10         Soda  ...  Mid-range product  Regularly busy  Regular days   
11         Soda  ...  Mid-range product  Regularly busy  Regular days   
12         Soda  ...  Mid-range product  Regularly busy  Busiest days   
13         Soda  ...  Mid-range product  Regularly busy  Busiest days   
14         Soda  ...  Mid-range product  Regularly busy  Regular days   

   busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0         Average orders        10      New customer  6.372222   Low spender   
1            Most orders        10      New customer  6.372222   Low spender   
2         Average orders        10      New customer  6.372222   Low spender   
3            Most orders        10      New customer  6.372222   Low spender   
4         Average orders        1

## 04. Exporting Data

In [34]:
# exporting data into pickle 
ords_prods_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))